src/dune_pkg/opam_solver.ml

```diff
@@ -758,6 +758,26 @@ module Solver_result = struct
     }
 end
 
+let reachable pkgs ~roots =
+  let seen = ref Package_name.Set.empty in
+  let rec loop = function
+    | [] -> ()
+    | pkg :: rest ->
+      if Package_name.Set.mem !seen pkg
+      then loop rest
+      else (
+        seen := Package_name.Set.add !seen pkg;
+        let deps =
+          match Package_name.Map.find pkgs pkg with
+          | None -> []
+          | Some deps -> deps
+        in
+        loop (List.rev_append deps rest))
```

Comment:
What's the point of reversing the dependencies before pre-pending them to the list of names to expand? Is it just that it's more efficient that using `@` and has no impact on the correctness of this search?

Reply:
It's for performance, though it can be serious. `@` reverses the list on every application, therefore it requires traversing the entire list.